graphrag/index/workflows/v1/subflows/create_final_communities.py

```diff
@@ -0,0 +1,181 @@
+    )
+
+    combined_clusters = _filter_clusters(concatenated_clusters)
+
+    cluster_relationships = aggregate_df(
+        combined_clusters,
+        aggregations=[
+            {
+                "column": "id_2",  # this is the id of the edge from the join steps above
+                "to": "relationship_ids",
+                "operation": "array_agg_distinct",
+            },
+            {
+                "column": "source_id_1",
+                "to": "text_unit_ids",
+                "operation": "array_agg_distinct",
+            },
+        ],
+        groupby=[
+            "cluster",
+            "level_1",  # level_1 is the left side of the join
+        ],
+    )
+
+    all_clusters = aggregate_df(
+        graph_nodes,
+        aggregations=[{"column": "cluster", "to": "id", "operation": "any"}],
+        groupby=["cluster", "level"],
+    )
+
+    joined = all_clusters.merge(
+        cluster_relationships,
+        left_on="id",
+        right_on="cluster",
+        how="inner",
+        suffixes=cast(Suffixes, ["_1", "_2"]),
+    )
+
+    filtered = _filter_secondary(joined)
+
+    titled = _create_title(filtered)
+
+    titled["raw_community"] = titled[
+        "id"
+    ]  # TODO: Rodrigo says "raw_community" is temporary
+    return create_verb_result(
+        cast(
+            Table,
+            titled[
+                [
+                    "id",
+                    "title",
+                    "level",
+                    "raw_community",
+                    "relationship_ids",
+                    "text_unit_ids",
+                ]
+            ],
+        )
+    )
+
+
+def _filter_clusters(table: pd.DataFrame) -> pd.DataFrame:
+    # level_1 is the left side of the join
+    # level_2 is the right side of the join
+    args = FilterArgs(
+        column="level_1",
+        criteria=[
+            Criterion(
+                type=FilterCompareType.Column,
+                operator=StringComparisonOperator.Equals,
+                value="level_2",
+            ),
+        ],
+    )
+    filter_index = filter_df(table, args)
+    sub_idx = filter_index == True  # noqa: E712
+    idx = filter_index[sub_idx].index  # type: ignore
+    return cast(pd.DataFrame, table[table.index.isin(idx)].reset_index(drop=True))
+
+
+# we've filtered on the left, now we filter on the right
+def _filter_secondary(table: pd.DataFrame) -> pd.DataFrame:
+    args = FilterArgs(
+        column="level",
+        criteria=[
+            Criterion(
+                type=FilterCompareType.Column,
+                operator=StringComparisonOperator.Equals,
+                value="level_1",
+            ),
+        ],
+    )
+    filter_index = filter_df(table, args)
+    sub_idx = filter_index == True  # noqa: E712
+    idx = filter_index[sub_idx].index  # type: ignore
+    return cast(pd.DataFrame, table[table.index.isin(idx)].reset_index(drop=True))
+
+
+def _create_title(table: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Please correct me if I'm wrong but pretty sure we can replace this and it aux functions by doing this:

```python
def _create_title(table: pd.DataFrame) -> pd.DataFrame:
    # Directly concatenate the "Community" string with the "id" column, handling any necessary type conversion
    table["title"] = "Community" + table["id"].astype(str)

    return table
   ```